src/table/system.rs

```diff
@@ -1,66 +1,224 @@
+use super::boot::{BootServices, MemoryMapIter, MemoryMapKey};
+use super::runtime::RuntimeServices;
 use super::{cfg, Header, Revision};
 use core::slice;
 use crate::proto::console::text;
-use crate::{CStr16, Char16, Handle};
+use crate::{CStr16, Char16, Handle, Result, ResultExt, Status};
 
-/// The system table entry points for accessing the core UEFI system functionality.
-#[repr(C)]
-pub struct SystemTable {
-    header: Header,
-    /// Null-terminated string representing the firmware's vendor.
-    fw_vendor: *const Char16,
-    /// Revision of the UEFI specification the firmware conforms to.
-    pub fw_revision: Revision,
-    stdin_handle: Handle,
-    stdin: *mut text::Input,
-    stdout_handle: Handle,
-    stdout: *mut text::Output,
-    stderr_handle: Handle,
-    stderr: *mut text::Output,
-    /// Runtime services table.
-    pub runtime: &'static super::runtime::RuntimeServices,
-    /// Boot services table.
-    pub boot: &'static super::boot::BootServices,
-    /// Number of entires in the configuration table.
-    nr_cfg: usize,
-    /// Pointer to beginning of the array.
-    cfg_table: *mut cfg::ConfigTableEntry,
-}
+/// Boot-time version of the UEFI System Table
+///
+/// This is the view of the UEFI System Table that an UEFI application is
+/// provided with initially. It enables calling all UEFI services, including
+/// so-called boot services, and exchanging information with the standard input,
+/// output and error streams.
+///
+/// Once an UEFI OS loader is ready to take over the system, it can call
+/// `exit_boot_services` to terminate the UEFI boot services. This will consume
+/// the BootSystemTable (enabling the Rust borrow checker to tell you about any
+/// boot service handle that you forgot about) and give you a RuntimeSystemTable
+/// which is more appropriate for post-boot usage.
+#[repr(transparent)]
+pub struct BootSystemTable(&'static SystemTable);
 
 // This is unsafe, but it's the best solution we have from now.
 #[allow(clippy::mut_from_ref)]
-impl SystemTable {
+impl BootSystemTable {
```

Comment:
I think we can use traits here to avoid duplicating code. How about:

```rust
// "marker" structures
struct Boot;
struct Runtime;

// common "marker" trait
trait SystemTableType {}

impl SystemTableType for Boot {}
impl SystemTableType for Runtime {}

struct SystemTable<T: SystemTableType> {
    /* ... structure fields ... */
    // needed to make compiler happy
    _marker: PhantomData<T>,
}

impl<T: SystemTableType> SystemTable<T> {
    fn firmware_revision(&self) {}
}

impl SystemTable<Boot> {
    fn boot_services(&self) {}
}
```

Seems to [do its job](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=35e1aa93ffe395eabaf9c25240d98eae)

Reply:
Excellent idea! I adopted that in the latest commit.